### PR TITLE
Some comments and documentation for name resolution crate

### PR DESCRIPTION
**Pull request by vorner**
Hello

I'm trying to get a grasp of how the name resolution crate works, as part of helping with https://github.com/rust-lang-nursery/rustc-guide/issues/16. Not that I'd be succeeding much, but as I was reading the code, I started to put some notes into it, to help me understand.

I guess I didn't get very far yet, but I'd like to share what I have, in case it might be useful for someone else. I hope these are correct (even if incomplete), but I'll be glad for a fast check in case I put something misleading there.

### CODE REVIEW

**Comment by pietroalbini** [comment]
Thanks for this PR @vorner! We'll make sure someone from @rust-lang/compiler reviews this PR!

(@pnkfelix, ping from triage!)

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -662,6 +670,7 @@ impl<'tcx> Visitor<'tcx> for UsePlacementFinder {
     }
 }
 
+// This thing walks the whole crate in DFS manner, visiting each item, resolving names as it goes.
```
Comment (oli-obk) [inline src/librustc_resolve/lib.rs]: You can use doc comments. They'll show up on the impl in the docs.

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -864,6 +875,16 @@ enum RibKind<'a> {
...
+///
+/// A rib represents a scope names can live in. Note that these appear in many places, not just
+/// around braces. At any place where the list of accessible names (of the given namespace)
+/// changes, a new rib is put onto a stack. This may be, for example, a `let` statement (because it
+/// introduces variables), a macro, etc.
+///
+/// Different [rib kinds](enum.RibKind) are transparent for different names.
+///
+/// The resolution keeps a separate stack of ribs as it traverses the AST for each namespace. When
+/// resolving, the name is looked from inside out.
```
Comment (oli-obk) [inline src/librustc_resolve/lib.rs]: nit: `looked up`

**Comment by petrochenkov** [comment]
Sorry, missed this. Will review.

File: src/librustc_resolve/lib.rs [hunk 3]
```diff
@@ -1122,7 +1152,8 @@ impl<'a> NameBinding<'a> {
         resolver.get_macro(self.def_ignoring_ambiguity())
     }
 
-    // We sometimes need to treat variants as `pub` for backwards compatibility
+    // We sometimes need to treat variants as `pub` for backwards compatibility (eg. in enum
+    // variants)
```
Comment (oli-obk) [inline src/librustc_resolve/lib.rs]: This change doesn't seem to clarify more. `variants` are already mentioned

File: src/librustc_resolve/lib.rs [hunk 4]
```diff
@@ -1404,10 +1441,12 @@ impl<'a, 'b: 'a> ty::DefIdTree for &'a Resolver<'b> {
         match id.krate {
             LOCAL_CRATE => self.definitions.def_key(id.index).parent,
             _ => self.cstore.def_key(id).parent,
-        }.map(|index| DefId { index: index, ..id })
+        }.map(|index| DefId { index, ..id })
     }
 }
 
+// This is the interface through which the rest of the compiler asks about name resolution after
+// the whole AST has been indexed.
```
Comment (oli-obk) [inline src/librustc_resolve/lib.rs]: again, this can be a doc comment

**Review by oli-obk — CHANGES_REQUESTED** [review]

**Comment by vorner** [comment]
Good to know I can put a comment on trait implementation, I always thought it shows the description of the trait.

Should I squash the fixup once it looks good?

**Review by oli-obk — APPROVED** [review]

File: src/librustc_resolve/lib.rs [hunk 5]
```diff
@@ -909,7 +935,11 @@ enum PathResult<'a> {
 }
 
 enum ModuleKind {
+    /// Inline `mod something { ... }`.
```
Comment (jseyfried) [inline src/librustc_resolve/lib.rs]: `ModuleKind::Block` is an anonymous module, i.e. a block:
```rust
fn main() {
    fn f() {} // (1)
    {
        f(); // This resolves to (2)
        fn f() {} (2)
    }
}
```

Whether a module is inline or not has no effect on name resolution.

File: src/librustc_resolve/lib.rs [hunk 6]
```diff
@@ -909,7 +935,11 @@ enum PathResult<'a> {
...
     Block(NodeId),
+    /// Module from another file.
+    ///
+    /// Also called a normal module in the following code.
```
Comment (jseyfried) [inline src/librustc_resolve/lib.rs]: A normal module is a `mod` module (inline or not). `ModuleKind::Def` could also be a trait or an enum.

File: src/librustc_resolve/lib.rs [hunk 7]
```diff
@@ -1359,6 +1394,7 @@ pub struct Resolver<'a> {
     injected_crate: Option<Module<'a>>,
 }
 
+/// Nothing really interesting here, it just provides memory for the rest of the crate.
```
Comment (jseyfried) [inline src/librustc_resolve/lib.rs]: nit: memory -> arenas

File: src/librustc_resolve/lib.rs [hunk 8]
```diff
@@ -1404,10 +1440,12 @@ impl<'a, 'b: 'a> ty::DefIdTree for &'a Resolver<'b> {
         match id.krate {
             LOCAL_CRATE => self.definitions.def_key(id.index).parent,
             _ => self.cstore.def_key(id).parent,
-        }.map(|index| DefId { index: index, ..id })
+        }.map(|index| DefId { index, ..id })
     }
 }
 
+/// This is the interface through which the rest of the compiler asks about name resolution after
+/// the whole AST has been indexed.
```
Comment (jseyfried) [inline src/librustc_resolve/lib.rs]: `hir::lowering::Resolver` is used to resolve HIR paths generated due to desugaring during AST->HIR lowering, and to look up the resolution AST paths during lowering (`get_resolution`). After lowering, resolution information is embedded in the HIR; this is what the rest of the compiler uses. The `Resolver` is destroyed shortly after lowering.

File: src/librustc_resolve/lib.rs [hunk 9]
```diff
@@ -864,6 +875,16 @@ enum RibKind<'a> {
 }
 
 /// One local scope.
+///
+/// A rib represents a scope names can live in. Note that these appear in many places, not just
+/// around braces. At any place where the list of accessible names (of the given namespace)
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: It's probably better to say it more explicitly - the list of accessible names changes or new restrictions on name accessibility are added.

**Comment by petrochenkov** [comment]
r=me after addressing the comments

**Comment by shepmaster** [comment]
Ping from triage @vorner ! You have pending feedback to address; will you be able to address it soon?

File: src/librustc_resolve/lib.rs (hunk 7 above)
Comment (vorner) [reply src/librustc_resolve/lib.rs]: I think it is obvious it contains arenas, my point was that these arenas are there to allocate memory. Should I still change it?

**Comment by vorner** [comment]
OK, added one more fixup commit.

**Pull request merged**